app/assets/javascripts/components/amounttable.js

```diff
@@ -0,0 +1,313 @@
+        this.makeTooltip = tooltip;
+        if (tooltip !== null) {
+            if (tooltipID === null) {
+                this.createTooltip();
+            } else {
+                this.tooltip = d3.select(tooltipID);
+            }
+        } else {
+            this.tooltip = null;
+        }
+    }
+
+    /**
+     * Create a tooltip element of class "tip" do display tooltips
+     */
+    createTooltip() {
+        this.tooltip = d3.select("body")
+            .append("div")
+            .attr("class", "tip")
+            .style("position", "absolute")
+            .style("z-index", "10")
+            .style("visibility", "hidden")
+            .style("background-color", "white")
+            .style("padding", "2px")
+            .style("border", "1px solid #dddddd")
+            .style("border-radius", "3px;");
+    }
+
+    /**
+     * Make the table visualization
+     */
+    draw() {
+        const [thead, tbody] = this.buildTable();
+        this.buildHeader(thead);
+        if (this.data.length != 0) {
+            this.buildRow(tbody);
+        } else {
+            // fallback in case there is no data
+            tbody.append("tr").append("td")
+                .attr("colspan", this.header.length)
+                .style("text-align", "center")
+                .text("No data");
+        }
+    }
+
+    /**
+     * Create a table in the target element (without creating it first)
+     * @return {d3.selection}  Table head and table body
+     */
+    buildTable() {
+        if (this.table === null) {
+            this.table = this.el.append("table").attr("class", "table table-condensed table-amounttable");
+        }
+        this.table.html(); // empty the tables HTML
+        const thead = this.table.append("thead");
+        const tbody = this.table.append("tbody");
+        return [thead, tbody];
+    }
+
+    /**
+     * Make the table header
+     * @param {d3.selection} thead  table header
+     */
+    buildHeader(thead) {
+        let headerCells = thead.append("tr")
+            .selectAll("th")
+            .data(this.header)
+            .enter()
+            .append("th")
+            .attr("scope", "col")
+            .text(d => d);
+        let lastCell = d3.select(headerCells[0][headerCells.size()-1]);
+        lastCell
+            .append("button")
+            .classed("btn btn-default btn-xs btn-animate amounttable-download", true)
+            .html("<span class=\"glyphicon glyphicon-download down\"></span> Save CSV")
+            .on("click", ()=>this.downloadCSV());
+    }
+
+    /**
+     * Add rows to the table tbody
+     * @param {d3.selection} tbody table body
+     */
+    buildRow(tbody) {
+        let rows = tbody.selectAll("tr")
+            .data(this.collapsed ? this.data.slice(0, this.limit): this.data);
+
+        rows.exit().remove(); // remove rows that are no longer needed
+
+        let row = rows.enter().append("tr");
+        for (let colSpec of this.contents) {
```

Comment:
When using for ... of loops, use `const` over `let` if the variable isn't reassigned in the loop block. Change on several places in this file.

_[Original comment](https://github.ugent.be/unipept/unipept/pull/625#discussion_r13717) by @bmesuere on Mon Feb 19 2018 at 13:54._